File: Api/Companies/iCompanyPortal.Api.Companies.Client/CompanyInvitationsClient.cs
﻿using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace iCompanyPortal.Api.Companies.Client
{
    public class CompanyInvitationsClient : ICompanyInvitationsClient
    {
        private readonly HttpClient http;

        public CompanyInvitationsClient(HttpClient http)
        {
            this.http = http;
        }

        public async Task<CompanyInvitation[]> GetInvitationsAsync(int companyId)
        {
            return await http.GetAsync<CompanyInvitation[]>($"/{companyId}/all");
        }

        public async Task<CompanyInvitation> GetAsync(Guid token)
        {
            return await http.GetAsync<CompanyInvitation>($"/{token}");
        }

        public async Task InviteAsync(int companyId, string responseUrl, CompanyInvitationRequest request)
        {
            await http.PostAsync($"/{companyId}/{WebUtility.UrlEncode(responseUrl)}", request);
        }

        public async Task RespondAsync(Guid token, bool accepted)
        {
            await http.PutAsync($"/{token}/{accepted}");
        }

        public async Task<bool> ActivateAsync(string email, int userId)
        {
            return await http.PutAsync<bool>($"/{email}/{userId}/activate");
        }

        public async Task DeleteAsync(Guid token)
        {
            await http.DeleteAsync($"/{token}");
        }
    }
}

File: App/icompanyportal/src/app/services/api-clients/companies-api/company-invitations-client.ts
import { Injectable, Inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { COMPANIES_API_ENDPOINT } from './companies-api-endpoint';
import { CompanyInvitation } from './company-invitation';
import { CompanyInvitationRequest } from './company-invitation-request';

@Injectable({
  providedIn: 'root'
})
export class CompanyInvitationsClient {
  constructor(private http: HttpClient, @Inject(COMPANIES_API_ENDPOINT) private endpoint: string) {}

  getInvitations(companyId: number) {
    return this.http.get<CompanyInvitation[]>(`${this.endpoint}/invitations/${companyId}/all`);
  }

  get(token: string) {
    return this.http.get<CompanyInvitation>(`${this.endpoint}/invitations/${token}`);
  }

  invite(companyId: number, responseUrl: string, request: CompanyInvitationRequest) {
    return this.http.post<void>(`${this.endpoint}/invitations/${companyId}/${encodeURIComponent(responseUrl)}`, request);
  }

  respond(token: string, accepted: boolean) {
    return this.http.put<void>(`${this.endpoint}/invitations/${token}/${accepted}`, undefined);
  }

  activate(email: string, userId: number) {
    return this.http.put<void>(`${this.endpoint}/invitations/${encodeURIComponent(email)}/${userId}/activate`, undefined);
  }

  delete(token: string) {
    return this.http.delete<void>(`${this.endpoint}/invitations/${token}`);
  }
}
